Currency Converter, 1st iterration

itapaullus opened the pull request:
Нужно обсудить тесты, кажется они некорректно из под мавена запускаются у меня...

Commit:
Currency exchange console app

kegisa commented on pom.xml:
```diff
@@ -0,0 +1,32 @@
+        </dependency>
+    </dependencies>
+    <build>
+        <testSourceDirectory>src/test/java/implementations</testSourceDirectory>
```
давай пока без тестов

Commit:
add class to console testing

kegisa commented on src/main/java/ConsoleCurrencyConverter.java:
```diff
@@ -0,0 +1,52 @@
+            if (choise.equalsIgnoreCase("Q")) {
+                break;
+            } else if (choise.equalsIgnoreCase("C")) {
+                System.out.print("Enter currency code to convert from: ");
```
Лучше логику вынести в какой то отдельный класс, иначе не получится заменять интерфейсы взаимодействия с клиентом.

kegisa replied:
а лучше где то на более низком уровне его заменить на свое исключение

kegisa commented on src/main/java/ConsoleCurrencyConverter.java:
```diff
@@ -0,0 +1,52 @@
+                try {
+                    double new_amount = converter.convert(curFrom, curTo, amount);
+                    System.out.println("Your converted amount: " + new_amount + " " + curTo);
+                } catch (IllegalArgumentException e) {
```
Разве других исключений не может быть?

kegisa commented on src/main/java/ConsoleCurrencyConverter.java:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    public static void main(String[] args) throws IOException {
```
Почему мейн в этом же классе?

kegisa replied:
IOException тогда надо бы отловить

kegisa commented on src/main/java/ConsoleCurrencyConverter.java:
```diff
@@ -0,0 +1,52 @@
+                    System.out.println("Currency pair " + curFrom + "/" + curTo + " doesn't exists");
+                }
+            }
+            {
```
что то уехало

kegisa commented on src/main/java/implementations/JsonExchangeRateProvider.java:
```diff
@@ -0,0 +1,33 @@
+import model.CurrencyPair;
+
+import interfaces.ExchangeRateProvider;
+import org.json.*;
```
идею настрой на разворачивание всех импортов

kegisa commented on src/main/java/implementations/JsonExchangeRateProvider.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public Map<CurrencyPair, Double> getExchangeRates() throws IOException {
+        byte[] jsonData = Files.readAllBytes(Paths.get(filename));
```
точнго нужно это каждый вызов метода делать?

kegisa commented on src/main/java/implementations/SimpleCurrencyConverter.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Map;
+
+public class SimpleCurrencyConverter implements CurrencyConverter {
+    private final Map<CurrencyPair, Double> rates;
```
врядли конвертер должен хранить в себе курсы валют, это неплохо бы вынести в класск который эти валюты получает

kegisa commented on src/main/java/model/CurrencyPair.java:
```diff
@@ -0,0 +1,31 @@
+
+public class CurrencyPair {
+
+    public String curFrom;
```
private

kegisa commented on src/main/java/interfaces/CurrencyConverter.java:
```diff
@@ -0,0 +1,7 @@
+import java.io.IOException;
+
+public interface CurrencyConverter {
+    double convert (String curFrom, String curTo, Double Amount) throws IOException;
```
конвертер должен выбрасывать исключение связанные с вводом и выводом?

kegisa commented on src/main/java/ConsoleCurrencyConverter.java:
```diff
@@ -0,0 +1,52 @@
+                System.out.print("Enter amount to convert: ");
+                Double amount = console.nextDouble();
+                try {
+                    double new_amount = converter.convert(curFrom, curTo, amount);
```
camelCase

Commit:
refactoring after review

Commit:
bugfixes..

kegisa commented on src/main/java/ConverterApp.java:
```diff
@@ -0,0 +1,30 @@
+        CurrencyConverter converter = new SimpleCurrencyConverter(new JsonExchangeRateProvider("src/main/resources/rates/rates.json"));
+        UserInterface ui = new ConsoleUserInterface();
+        while (true) {
+            ui.ShowMenu();
```
методы с малкнькой буквы пишутся

kegisa commented on src/main/java/ConverterApp.java:
```diff
@@ -0,0 +1,30 @@
+        while (true) {
+            ui.ShowMenu();
+            Action action = ui.getAction();
+            if (action.equals(Action.Quit)) {
```
enum в апперкейсе лучше писать

itapaullus replied:
+

kegisa commented on src/main/java/ConverterApp.java:
```diff
@@ -0,0 +1,30 @@
+                break;
+            } else if (action.equals(Action.ExchangeCurrency)) {
+                try {
+                    System.out.println(converter.convert(
```
если изменим интерфейс взаимодействия ты все равно выведешь в консоль результат

kegisa commented on src/main/java/converter/SimpleCurrencyConverter.java:
```diff
@@ -0,0 +1,21 @@
+import rateprovider.ExchangeRateProvider;
+
+public class SimpleCurrencyConverter implements CurrencyConverter {
+    private final ExchangeRateProvider provider;
```
лишняя зависимость для конвертера, он может обойтись без этого, просто принимать на вход метода сразу валюту нужную и просто переводить, а  не вытаскивать данные

kegisa commented on src/main/java/converter/SimpleCurrencyConverter.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    @Override
+    public double convert(String curFrom, String curTo, Double amount) throws IllegalArgumentException{
```
врядли конвертер должен выбрасывать illegalargument он же просто считает, когда уберешь зависимость эти побочки тоже уйдут.

kegisa commented on src/main/java/rateprovider/JsonExchangeRateProvider.java:
```diff
@@ -0,0 +1,51 @@
+
+public class JsonExchangeRateProvider implements ExchangeRateProvider {
+    private final String filename;
+    private CurrencyRateRepository rates;
```
вот это по моему лишнее, но в целом ОК

Commit:
убраны зависимости между CurrencyConverter и ExchangeRateProvider.
Валюта убрана в enum
Все реализации собираются в App и передаются в ConverterApp
Все консольные операции убраны в ConsoleUserInterface
Добавлена обработка ошибок неверных валют, суммы конвертации, действий в ui

kegisa commented on src/main/java/model/Currency.java:
```diff
@@ -0,0 +1,38 @@
+    RUR("RUR");
+    // Add more currencies here
+
+    private static final Map<String, Currency> currencyMap = new HashMap<>();
```
Зачем в enum логику добавлять?

kegisa commented on src/main/java/rateprovider/JsonExchangeRateProvider.java:
```diff
@@ -0,0 +1,53 @@
+                        new CurrencyPair(curFrom, curTo),
+                        rate.getDouble("value"));
+            }
+            this.rates = new InMemoryCurrencyRateRepository(exchangeRates);
```
хардкод реализации, я бы вообще эту прослойку убрал чтобы не путаться, сделай просто здесь мапу в этом классе, первый раз когда запросят валюту заполни эту мапуиз файла, далее бери из мапы, если вдруг в мапе не найдено надо снова сходить в файл перезаполнрить мапу и попрбовать найти, если ненайдено тогда исключение

kegisa commented on src/main/java/ui/Action.java:
```diff
@@ -0,0 +1,35 @@
+    QUIT("Q");
+    private static final Map<String, Action> actionMap = new HashMap<>();
+
+    static {
```
я не понимаею зачем ты в enum этот код пишешь?

Commit:
Убран RateRepository
Убрана логика из enum
Убран Enum Currency
Map с курсами вынесен в ExchangeRateProvider, заполняется при обращении к курсам, когда валютная пара не найдена
